### ISSUE TITLE
Discussion: Target-typed implicit array creation expression `new[]` 

### ISSUE THREAD

**Issue by alrz** [dotnet/csharplang#2701]
With introduction of `new()` expression (#100), the following:
```cs
IEnumerable<KeyValuePair<string, string>> Headers = new[]
{
     new KeyValuePair<string, string>("Foo", foo),
     new KeyValuePair<string, string>("Bar", bar),
}
```
can be simplified to:
```cs
IEnumerable<KeyValuePair<string, string>> Headers = new KeyValuePair<string, string>[]
{
     new("Foo", foo),
     new("Bar", bar),
}
```
but you still need to repeat the type following the field/property initializer. Closest you can get is something like: 
```cs
IEnumerable<KeyValuePair<string, string>> Headers = new[]
{
     new KeyValuePair<string, string>("Foo", foo),
     new("Bar", bar),
}
```
For the sake of completeness, I'd suggest to also make `new[]` a target-typed expression,
```cs
IEnumerable<KeyValuePair<string, string>> Headers = new[]
{
     new("Foo", foo),
     new("Bar", bar),
}
```
The mechanics of such feature is similar to #2389 or #2460 where there are two source of type inference, the common type and the target-type.

Note: A list/dictionary literal feature would supersede this proposal because they are most likely target-typed from get-go.

**Comment by alrz** [comment]
As with how this feature would interact with overload resolution there's some room for tweaks,


```cs
void M(List<int> list) {}
void M(IEnumerable<int>  list) {}
```
If `new[]` would be convertible to any type just like `new()`, both this invocations would be ambiguous,
```cs
M(new() { }); // could bind to first method
M(new[] { }); // could bind to second method
```
While this is good for breaking change predictability, it narrows down where the feature can be used.

--

For an untyped `new[]` we could only accept *arrays* AND *array generic interfaces* as the target type and yield no conversion otherwise. This would be similar to lambas which are only convertible to delegates.

For an untyped `new()` we could only accept *constructible types* or more loosely *non-array types* instead of *any* type. 

This distinction would eliminate the candidates that are not compatible with the argument from get-go, since we can say if we want to create an array or a non-array type via `new[]` or `new()`.

**Comment by lcsondes** [comment]
Would it not work better to relax the array initializer syntax to allow `{{a,b},{c,d}}` instead of `{new(a,b),new(c,d)}` for the KeyValuePair example?

**Comment by lcsondes** [comment]
I would like if this also included `int[] x = new[1];`, not just `new[]{...}`.

**Comment by CyrusNajmabadi** [comment]
I think this proposal can effectively be subsumed into [collection literals](https://github.com/dotnet/csharplang/issues/5354). 

You would be able to do the following:

```c#
KeyValuePair<string, string>[] array = [new("key1", value1), new("key2", value2)];
```

That solves the array part.  KVP is also very special to collection literals so the above could also be shortened to:

```C#
KeyValuePair<string, string>[] array = ["key1": value1), "key2": value2];
```

> Note: A list/dictionary literal feature would not supersede this proposal because in above examples there's no concrete collection type to infer.

We support collection/dictionary literals even the absence of a concrete collection type :)

**Comment by alrz** [comment]
Right, I think it will also work with IEnumerable but only as long as there is a natural type.
```cs
IEnumerable<T> x = [value]; // ok; but the target-type (T) won't affect type inference in any ways here.
```
```cs
IEnumerable<object> x = []; // error; no natural type.
IEnumerable<bool?> x = [true, null]; // error; the most-common-type is `bool` and `List<bool>` fails to assign.
```
Relates to https://github.com/dotnet/roslyn/issues/37950

**Comment by CyrusNajmabadi** [comment]
All of the above would be legal @alrz

**Comment by alrz** [comment]
Great! Though I didn't find anything about a null-aware common type.. does that last example work with `var`?

**Comment by CyrusNajmabadi** [comment]
The last example will work with 'var' *if* 'best common type' is willing to infer `bool?` for `true` and `null`.  So we would be gated on that.

**Comment by RikkiGibson** [comment]
> _if_ 'best common type' is willing to infer `bool?` for `true` and `null`.

Unfortunately, I think it is not. [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgPgAQEwEYCwAoBBmABM3AYVwG9NcL88EAWXAWQAoBKU8yjgNwgCdcAPXAF5cMAKYB3ANoBdUrgAuPAK5iANKOUAbLbgC+AbkzsOFbnwBmWiAHNhilWKMZTlc7gCe9q7dwB+B1VcEE0dZw49TD0gA=)

**Comment by CyrusNajmabadi** [comment]
@RikkiGibson right, that's what i'm saying.  We're gated on that being a supported concept for c#.  the same applies to existing collections today.

**Comment by 333fred** [comment]
When we didn't do that in C# 8, we accepted that it would likely never be done. https://github.com/dotnet/csharplang/issues/881

**Comment by audacode** [comment]
I opened a discussion thread here: https://github.com/dotnet/csharplang/discussions/7175
But closed it once I found this issue..

In short: It is great if this type of syntax will work in C# 12:
```cs
Student[] students = [new("John"),new("Jane")];
```
But we still should have a goal of implementing this feature on the original syntax for the sake of language consistency.
```cs
Student[] studentsB = new[] { new("John"), new("Jane") }; // This would ideally work as well.
```

**Comment by alrz** [comment]
>for the sake of language consistency

I wouldn't say it's only that, because the two would be different in natural type: `List<T>` versus `T[]`


The curious part is that a collection literal also falls short if you want to use type inference for elements:

```cs
// assuming there exist a natural type, maybe List<T>
var students = [new("John"),new("Jane")]; // ERROR
```


It would be nice to be able to do that more concisely, regardless of how the natural type comes to be:

```cs
var x = [] ConcereteElementType; // or something like that

List<IMyInterface> x = [new(1, 2), new(3,4)] ConcereteElementType;
```
Currently, making this "work" is kind of awkward: specify the type for one of the elements, if any.
